File: app/src/main/java/com/cookplan/recipe/edit/add_ingredients/EditRecipeIngredientsPresenterImpl.java
package com.cookplan.recipe.edit.add_ingredients;


import com.cookplan.models.CookPlanError;
import com.cookplan.models.Ingredient;
import com.cookplan.models.Recipe;
import com.cookplan.providers.IngredientProvider;
import com.cookplan.providers.impl.IngredientProviderImpl;

import java.util.List;

import io.reactivex.CompletableObserver;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by DariaEfimova on 21.03.17.
 */

public class EditRecipeIngredientsPresenterImpl implements EditRecipeIngredientsPresenter {

    private EditRecipeIngredientsView mainView;
    private Recipe recipe;

    private IngredientProvider dataProvider;
    private CompositeDisposable disposables;

    EditRecipeIngredientsPresenterImpl(EditRecipeIngredientsView mainView, Recipe recipe) {
        this.mainView = mainView;
        this.recipe = recipe;
        this.dataProvider = new IngredientProviderImpl();
        disposables = new CompositeDisposable();
    }

    @Override
    public void getAsyncIngredientList() {
        disposables.add(dataProvider.getIngredientListByRecipeId(recipe.getId())
                                .subscribeOn(Schedulers.io())
                                .observeOn(AndroidSchedulers.mainThread())
                                .subscribeWith(new DisposableObserver<List<Ingredient>>() {

                                    @Override
                                    public void onNext(List<Ingredient> ingredients) {
                                        if (mainView != null) {
                                            mainView.setIngredientList(ingredients);
                                        }
                                    }

                                    @Override
                                    public void onError(Throwable e) {
                                        if (mainView != null && e instanceof CookPlanError) {
                                            mainView.setErrorToast(e.getMessage());
                                        }
                                    }

                                    @Override
                                    public void onComplete() {

                                    }
                                }));
    }

    @Override
    public void removeIngredient(Ingredient ingredient) {
        dataProvider.removeIngredient(ingredient)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new CompletableObserver() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onComplete() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        if (mainView != null && e instanceof CookPlanError) {
                            mainView.setErrorToast(e.getMessage());
                        }
                    }
                });
    }

    @Override
    public void onStop() {
        disposables.clear();
    }
}

File: app/src/main/java/com/cookplan/recipe/edit/ingredients/EditRecipeIngredientsPresenterImpl.kt
package com.cookplan.recipe.edit.ingredients


import com.cookplan.models.Ingredient
import com.cookplan.models.Recipe
import com.cookplan.providers.IngredientProvider
import com.cookplan.providers.ProviderFactory
import io.reactivex.CompletableObserver
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.observers.DisposableObserver
import io.reactivex.schedulers.Schedulers

/**
 * Created by DariaEfimova on 21.03.17.
 */

class EditRecipeIngredientsPresenterImpl internal constructor(private val mainView: EditRecipeIngredientsView?, private val recipe: Recipe?) : EditRecipeIngredientsPresenter {

    private val dataProvider: IngredientProvider = ProviderFactory.Companion.ingredientProvider
    private val disposables: CompositeDisposable = CompositeDisposable()

    override fun getAsyncIngredientList() {
        disposables.add(dataProvider.getIngredientListByRecipeId(recipe?.id ?: String())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(object : DisposableObserver<List<Ingredient>>() {

                    override fun onNext(ingredients: List<Ingredient>) {
                        mainView?.setIngredientList(ingredients)
                    }

                    override fun onError(e: Throwable) {
                        mainView?.setErrorToast(e.message ?: String())
                    }

                    override fun onComplete() {

                    }
                }))
    }

    override fun removeIngredient(ingredient: Ingredient) {
        dataProvider.removeIngredient(ingredient)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : CompletableObserver {
                    override fun onSubscribe(d: Disposable) {

                    }

                    override fun onComplete() {

                    }

                    override fun onError(e: Throwable) {
                        mainView?.setErrorToast(e.message ?: String())
                    }
                })
    }

    override fun onDestroy() {
        disposables.clear()
    }
}